app/src/main/java/org/oppia/android/app/testing/AdministratorControlsFragmentTestActivity.kt

```diff
@@ -0,0 +1,66 @@
+package org.oppia.android.app.testing
+
+import android.content.Context
+import android.content.Intent
+import android.os.Bundle
+import org.oppia.android.app.activity.ActivityComponentImpl
+import org.oppia.android.app.administratorcontrols.LogoutDialogFragment
+import org.oppia.android.app.administratorcontrols.RouteToAppVersionListener
+import org.oppia.android.app.administratorcontrols.RouteToProfileListListener
+import org.oppia.android.app.administratorcontrols.ShowLogoutDialogListener
+import org.oppia.android.app.administratorcontrols.appversion.AppVersionActivity
+import org.oppia.android.app.drawer.NAVIGATION_PROFILE_ID_ARGUMENT_KEY
+import org.oppia.android.app.settings.profile.ProfileListActivity
+import org.oppia.android.app.testing.activity.TestActivity
+import org.oppia.android.app.translation.AppLanguageResourceHandler
+import javax.inject.Inject
+
+/** Activity [AdministratorControlsFragmentTestActivity] that allows user to change admin controls. */
+class AdministratorControlsFragmentTestActivity :
+  TestActivity(),
+  RouteToProfileListListener,
+  RouteToAppVersionListener,
+  ShowLogoutDialogListener {
+  @Inject
+  lateinit var administratorControlsFragmentTestActivityPresenter:
+    AdministratorControlsFragmentTestActivityPresenter
+
+  @Inject
+  lateinit var resourceHandler: AppLanguageResourceHandler
+
+  override fun onCreate(savedInstanceState: Bundle?) {
+    super.onCreate(savedInstanceState)
+    (activityComponent as ActivityComponentImpl).inject(this)
+    administratorControlsFragmentTestActivityPresenter.handleOnCreate()
+  }
+
+  override fun routeToAppVersion() {
+    startActivity(AppVersionActivity.createAppVersionActivityIntent(this))
+  }
+
+  override fun routeToProfileList() {
+    startActivity(ProfileListActivity.createProfileListActivityIntent(this))
+  }
+
+  companion object {
+    /** Returns an [Intent] to start this activity. */
+    fun createAdministratorControlsFragmentTestActivityIntent(
+      context: Context,
+      profileId: Int?
+    ): Intent {
+      val intent = Intent(context, AdministratorControlsFragmentTestActivityPresenter::class.java)
+      intent.putExtra(NAVIGATION_PROFILE_ID_ARGUMENT_KEY, profileId)
+      return intent
+    }
+
+    /** Returns the argument key used to specify the user's internal profile ID. */
+    fun getIntentKey(): String {
+      return NAVIGATION_PROFILE_ID_ARGUMENT_KEY
+    }
+  }
+
+  override fun showLogoutDialog() {
```

Comment:
But it is already tested in both the activity and fragment

Reply:
Tests generally should be verifying the "public API" (i.e. observable behaviors) of their component under test.

For an activity, that's essentially the high-level user behaviors (especially those which go across fragments since individual fragment tests can't verify them). Showing the dialog is an example of this: the activity is responsible for showing the dialog under certain conditions. Since it's the activity's responsibility, we verify it in the activity's test suite.

The fragment, on the other hand, isn't responsible for _showing_ the dialog, but it _is_ responsible for notifying its host activity when it's time to show the dialog. Because this is an expected behavior of the fragment, we should test that. While verifying that the dialog also opens is technically verifying it by proxy, that's entirely contingent on the test activity behaving the same way as the production activity. My suggestion is to instead pass in a fake or mock ShowLogoutDialogListener that the test activity just calls here rather than doing anything, and then we can verify mor explicitly in the fragment's test suite that it calls the activity's ShowLogoutDialogListener when expected.

Does that help clarify?

Reply:
thanks ben this clears the whole thing I have moved tests related to these to activity and also removed the implementation and just overridden the methods.